borg/logger.py

```diff
@@ -84,5 +126,48 @@ def create_logger(name=None):
         logger = logging.getLogger(__name__)
 
     we use it to avoid errors and provide a more standard API.
+
+    We must create the logger lazily, because this is usually called from
+    module level (and thus executed at import time - BEFORE setup_logging()
+    was called). By doing it lazily we can do the setup first, we just have to
+    be careful not to call any logger methods before the setup_logging() call.
+    If you try, you'll get an exception.
     """
-    return logging.getLogger(name or find_parent_module())
+    class LazyLogger:
+        def __init__(self, name=None):
+            self.__name = name or find_parent_module()
+            self.__real_logger = None
+
+        @property
+        def __logger(self):
+            if self.__real_logger is None:
+                if not configured:
+                    raise Exception("tried to call a logger before setup_logging() was called")
+                self.__real_logger = logging.getLogger(self.__name)
+            return self.__real_logger
+
+        def setLevel(self, *args, **kw):
+            return self.__logger.setLevel(*args, **kw)
+
+        def log(self, *args, **kw):
+            return self.__logger.log(*args, **kw)
+
+        def exception(self, *args, **kw):
+            return self.__logger.exception(*args, **kw)
+
+        def debug(self, *args, **kw):
+            return self.__logger.debug(*args, **kw)
+
+        def info(self, *args, **kw):
+            return self.__logger.info(*args, **kw)
+
+        def warning(self, *args, **kw):
+            return self.__logger.warning(*args, **kw)
+
+        def error(self, *args, **kw):
+            return self.__logger.error(*args, **kw)
+
+        def critical(self, *args, **kw):
+            return self.__logger.critical(*args, **kw)
+
+    return LazyLogger(name)
```

Comment:
the thing is we control where loggers are created (in create_logger), that is the whole point. we can just intercept there and crash if the module wasn't configured properly. i don't think lazy loading is necessary in this case, since we do not create_logger before configuring it anywhere (and if we do, we can just crash, i think that's a fair API).

... but what do i know.